src/typeguard/_transformer.py

```diff
@@ -500,17 +499,23 @@ def __init__(
         self.target_lineno = target_lineno
 
     def generic_visit(self, node: AST) -> AST:
-        non_empty_list_fields = []
-        for field_name, val in iter_fields(node):
-            if isinstance(val, list) and len(val) > 0:
-                non_empty_list_fields.append(field_name)
+        has_non_empty_body_initially = (
+            hasattr(node, "body") and node.body and len(node.body) > 0
+        )
+        initial_type = type(node)
 
         node = super().generic_visit(node)
 
-        # Add `pass` to list fields that were optimised away
-        for field_name in non_empty_list_fields:
-            if not getattr(node, field_name, None):
-                setattr(node, field_name, [Pass()])
+        new_type = type(node)
+        if (
+            initial_type == new_type
+            and has_non_empty_body_initially
+            and hasattr(node, "body")
+            and (not node.body or len(node.body) == 0)
```

Comment:
```suggestion
            and getattr(node, "body", None)
```

Reply:
This one is a bit more difficult. If I remove `hasattr(node, "body")`, then `mypy` cannot infer that the `node` actually has a `body`.

Reply:
If you prefer, I can accept this change at the expense of either:
* Changing `node.body = ...` in the body of the if statement to a `setattr` equivalent with an annotation to prohibit `ruff` to changing back to normal assignment;
* Or adding a type ignore to pacify `mypy`.

Reply:
Then how about:
```
and hasattr(node, "body")
and not node.body
```
The `len(node.body) == 0` check is irrelevant because if the body was a sequence of length 0, it would already evaluate as falsy in the previous check.

Reply:
done